Hibernate custom mapping to user-defined type (Enum) fails

**Describe the problem**

Following [Enum Mappings with Hibernate](https://thorben-janssen.com/hibernate-enum-mappings/#customizedMapping) tutorial and a StackOverflow [post](https://stackoverflow.com/questions/27804069/hibernate-mapping-between-postgresql-enum-and-java-enum), I tried to create a Hibernate custom mapping to the database Enum type.

The `reviews` table is created in the database however I am unable to insert a row and get the following stack trace.
```
root@:26257/vy> show create reviews;
  table_name |                create_statement
-------------+-------------------------------------------------
  reviews    | CREATE TABLE public.reviews (
             |     id INT8 NOT NULL,
             |     message VARCHAR(255) NULL,
             |     rating public.rating NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (id ASC),
             |     FAMILY "primary" (id, message, rating)
             | )
(1 row)
```

```
Aug 24, 2020 10:17:34 PM org.hibernate.engine.jdbc.spi.SqlExceptionHelper logExceptions
WARN: SQL Error: 0, SQLState: XX000
Aug 24, 2020 10:17:34 PM org.hibernate.engine.jdbc.spi.SqlExceptionHelper logExceptions
ERROR: ERROR: internal error: error in argument for $2: unsupported OID 100053 with format code FormatText
  Detail: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase/encoding.go:699: DecodeOidDatum()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:334: execBind()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1512: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1370: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:510: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580: func1()
runtime/asm_amd64.s:1357: goexit()

  Hint: You have encountered an unexpected error.

Please check the public issue tracker to check whether this problem is
already tracked. If you cannot find it there, please report the error
with details by creating a new issue.

If you would rather not post publicly, please contact us directly
using the support form.

We appreciate your feedback.

Aug 24, 2020 10:17:34 PM org.hibernate.engine.jdbc.connections.internal.DriverManagerConnectionProviderImpl$PoolState stop
INFO: HHH10001008: Cleaning up connection pool [jdbc:postgresql://localhost:26257/vy?sslmode=disable]
Exception in thread "main" javax.persistence.PersistenceException: org.hibernate.exception.GenericJDBCException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1364)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:451)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3210)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2378)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:447)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:183)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$300(JdbcResourceLocalTransactionCoordinatorImpl.java:40)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:281)
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:101)
	at com.cockroachlabs.Enum.main(Enum.java:55)
Caused by: org.hibernate.exception.GenericJDBCException: could not execute statement
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:47)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3255)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3782)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:723)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:348)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:102)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1360)
	... 9 more
Caused by: org.postgresql.util.PSQLException: ERROR: internal error: error in argument for $2: unsupported OID 100053 with format code FormatText
  Detail: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase/encoding.go:699: DecodeOidDatum()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:334: execBind()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1512: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1370: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:510: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580: func1()
runtime/asm_amd64.s:1357: goexit()

  Hint: You have encountered an unexpected error.

Please check the public issue tracker to check whether this problem is
already tracked. If you cannot find it there, please report the error
with details by creating a new issue.

If you would rather not post publicly, please contact us directly
using the support form.

We appreciate your feedback.

	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2433)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2178)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:306)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:155)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:132)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
	... 20 more

FAILURE: Build failed with an exception.

```

**To Reproduce**

see [reproduction code](https://github.com/vy-ton/hello-world-java-hibernate/blob/enum-testing/src/main/java/com/cockroachlabs/Enum.java)

**Environment:**
 - Build commit ID [a337502248982b27eea1ac8f348d1c12c904ca57]
 - Client app [see here](https://github.com/vy-ton/hello-world-java-hibernate/blob/enum-testing/src/main/java/com/cockroachlabs/Enum.java)

--- a/pkg/sql/conn_executor_prepare.go
+++ b/pkg/sql/conn_executor_prepare.go
@@ -19,6 +19,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase"
 	"github.com/cockroachdb/cockroach/pkg/sql/querycache"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/fsm"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -46,6 +47,20 @@ func (ex *connExecutor) execPrepare(
 	} else {
 		// Deallocate the unnamed statement, if it exists.
 		ex.deletePreparedStmt(ctx, "")
+	}
+
+	// If we were provided any type hints, attempt to resolve any user defined
+	// type OIDs into types.T's.
+	if parseCmd.TypeHints != nil {
+		for i := range parseCmd.TypeHints {
+			if parseCmd.TypeHints[i] == nil && types.IsOIDUserDefinedType(parseCmd.RawTypeHints[i]) {
+				var err error
+				parseCmd.TypeHints[i], err = ex.planner.ResolveTypeByOID(ctx, parseCmd.RawTypeHints[i])
+				if err != nil {
+					return retErr(err)
+				}
+			}
+		}
 	}
 
 	ps, err := ex.addPreparedStmt(
@@ -331,7 +346,7 @@ func (ex *connExecutor) execBind(
 				// nil indicates a NULL argument value.
 				qargs[k] = tree.DNull
 			} else {
-				d, err := pgwirebase.DecodeOidDatum(ptCtx, t, qArgFormatCodes[i], arg)
+				d, err := pgwirebase.DecodeOidDatum(ctx, ptCtx, t, qArgFormatCodes[i], arg, &ex.planner)
 				if err != nil {
 					return retErr(pgerror.Wrapf(err, pgcode.ProtocolViolation,
 						"error in argument for %s", k))

--- a/pkg/sql/copy.go
+++ b/pkg/sql/copy.go
@@ -364,10 +364,12 @@ func (c *copyMachine) readBinaryTuple(ctx context.Context) error {
 			return errors.Newf("partial copy data row")
 		}
 		d, err := pgwirebase.DecodeOidDatum(
+			ctx,
 			c.parsingEvalCtx,
 			c.resultColumns[i].Typ.Oid(),
 			pgwirebase.FormatBinary,
 			data,
+			&c.p,
 		)
 		if err != nil {
 			return pgerror.Wrapf(err, pgcode.BadCopyFileFormat,

--- a/pkg/sql/pgwire/conn.go
+++ b/pkg/sql/pgwire/conn.go
@@ -782,6 +782,12 @@ func (c *conn) handleParse(
 			if t == 0 {
 				continue
 			}
+			// If the OID is user defined, then write nil into the type hints and let
+			// the consumer of the PrepareStmt resolve the types.
+			if types.IsOIDUserDefinedType(t) {
+				sqlTypeHints[i] = nil
+				continue
+			}
 			v, ok := types.OidToType[t]
 			if !ok {
 				err := pgwirebase.NewProtocolViolationErrorf("unknown oid type: %v", t)

--- a/pkg/sql/pgwire/encoding_test.go
+++ b/pkg/sql/pgwire/encoding_test.go
@@ -183,7 +183,7 @@ func TestEncodings(t *testing.T) {
 				pgwirebase.FormatText:   tc.TextAsBinary,
 				pgwirebase.FormatBinary: tc.Binary,
 			} {
-				d, err := pgwirebase.DecodeOidDatum(nil, tc.Oid, code, value)
+				d, err := pgwirebase.DecodeOidDatum(context.Background(), nil, tc.Oid, code, value, nil)
 				if err != nil {
 					t.Fatal(err)
 				}
@@ -241,7 +241,7 @@ func TestExoticNumericEncodings(t *testing.T) {
 	evalCtx := tree.MakeTestingEvalContext(nil)
 	for i, c := range testCases {
 		t.Run(fmt.Sprintf("%d_%s", i, c.Value), func(t *testing.T) {
-			d, err := pgwirebase.DecodeOidDatum(nil, oid.T_numeric, pgwirebase.FormatBinary, c.Encoding)
+			d, err := pgwirebase.DecodeOidDatum(context.Background(), nil, oid.T_numeric, pgwirebase.FormatBinary, c.Encoding, nil)
 			if err != nil {
 				t.Fatal(err)
 			}

--- a/pkg/sql/pgwire/pgwirebase/encoding.go
+++ b/pkg/sql/pgwire/pgwirebase/encoding.go
@@ -13,6 +13,7 @@ package pgwirebase
 import (
 	"bufio"
 	"bytes"
+	"context"
 	"encoding/binary"
 	"io"
 	"math"
@@ -215,9 +216,15 @@ func validateArrayDimensions(nDimensions int, nElements int) error {
 
 // DecodeOidDatum decodes bytes with specified Oid and format code into
 // a datum. If the ParseTimeContext is nil, reasonable defaults
-// will be applied.
+// will be applied. If res is nil, then user defined types are not attempted
+// to be resolved.
 func DecodeOidDatum(
-	ctx tree.ParseTimeContext, id oid.Oid, code FormatCode, b []byte,
+	ctx context.Context,
+	pCtx tree.ParseTimeContext,
+	id oid.Oid,
+	code FormatCode,
+	b []byte,
+	res tree.TypeReferenceResolver,
 ) (tree.Datum, error) {
 	switch code {
 	case FormatText:
@@ -283,19 +290,19 @@ func DecodeOidDatum(
 			}
 			return tree.NewDBytes(tree.DBytes(res)), nil
 		case oid.T_timestamp:
-			d, _, err := tree.ParseDTimestamp(ctx, string(b), time.Microsecond)
+			d, _, err := tree.ParseDTimestamp(pCtx, string(b), time.Microsecond)
 			if err != nil {
 				return nil, pgerror.Newf(pgcode.Syntax, "could not parse string %q as timestamp", b)
 			}
 			return d, nil
 		case oid.T_timestamptz:
-			d, _, err := tree.ParseDTimestampTZ(ctx, string(b), time.Microsecond)
+			d, _, err := tree.ParseDTimestampTZ(pCtx, string(b), time.Microsecond)
 			if err != nil {
 				return nil, pgerror.Newf(pgcode.Syntax, "could not parse string %q as timestamptz", b)
 			}
 			return d, nil
 		case oid.T_date:
-			d, _, err := tree.ParseDDate(ctx, string(b))
+			d, _, err := tree.ParseDDate(pCtx, string(b))
 			if err != nil {
 				return nil, pgerror.Newf(pgcode.Syntax, "could not parse string %q as date", b)
 			}
@@ -307,7 +314,7 @@ func DecodeOidDatum(
 			}
 			return d, nil
 		case oid.T_timetz:
-			d, _, err := tree.ParseDTimeTZ(ctx, string(b), time.Microsecond)
+			d, _, err := tree.ParseDTimeTZ(pCtx, string(b), time.Microsecond)
 			if err != nil {
 				return nil, pgerror.Newf(pgcode.Syntax, "could not parse string %q as timetz", b)
 			}
@@ -668,7 +675,7 @@ func DecodeOidDatum(
 		default:
 			if _, ok := types.ArrayOids[id]; ok {
 				innerOid := types.OidToType[id].ArrayContents().Oid()
-				return decodeBinaryArray(ctx, innerOid, b, code)
+				return decodeBinaryArray(ctx, pCtx, innerOid, b, code, res)
 			}
 		}
 	default:
@@ -695,10 +702,32 @@ func DecodeOidDatum(
 			return nil, err
 		}
 		return tree.NewDName(string(b)), nil
-	default:
-		return nil, errors.AssertionFailedf(
-			"unsupported OID %v with format code %s", errors.Safe(id), errors.Safe(code))
 	}
+
+	// Finally, try to resolve the type's oid as a user defined type if a resolver
+	// was provided.
+	if res != nil {
+		typ, err := res.ResolveTypeByOID(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		switch typ.Family() {
+		case types.EnumFamily:
+			if code != FormatText {
+				return nil, pgerror.Newf(pgcode.Syntax, "expected FormatText for ENUM value encoding")
+			}
+			if err := validateStringBytes(b); err != nil {
+				return nil, err
+			}
+			return tree.MakeDEnumFromLogicalRepresentation(typ, string(b))
+		default:
+			return nil, errors.AssertionFailedf("unsupported user defined type family %s", typ.Family().String())
+		}
+	}
+
+	// Fallthrough case.
+	return nil, errors.AssertionFailedf(
+		"unsupported OID %v with format code %s", errors.Safe(id), errors.Safe(code))
 }
 
 // Values which are going to be converted to strings (STRING and NAME) need to
@@ -797,7 +826,12 @@ func pgBinaryToIPAddr(b []byte) (ipaddr.IPAddr, error) {
 }
 
 func decodeBinaryArray(
-	ctx tree.ParseTimeContext, elemOid oid.Oid, b []byte, code FormatCode,
+	ctx context.Context,
+	pCtx tree.ParseTimeContext,
+	elemOid oid.Oid,
+	b []byte,
+	code FormatCode,
+	res tree.TypeReferenceResolver,
 ) (tree.Datum, error) {
 	var hdr struct {
 		Ndims int32
@@ -842,7 +876,7 @@ func decodeBinaryArray(
 			continue
 		}
 		buf := r.Next(int(vlen))
-		elem, err := DecodeOidDatum(ctx, elemOid, code, buf)
+		elem, err := DecodeOidDatum(ctx, pCtx, elemOid, code, buf, res)
 		if err != nil {
 			return nil, err
 		}

--- a/pkg/sql/pgwire/pgwirebase/fuzz.go
+++ b/pkg/sql/pgwire/pgwirebase/fuzz.go
@@ -13,6 +13,8 @@
 package pgwirebase
 
 import (
+	"context"
+
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -40,7 +42,7 @@ func FuzzDecodeOidDatum(data []byte) int {
 	code := FormatCode(data[0]) % (FormatBinary + 1)
 	b := data[2:]
 
-	_, err := DecodeOidDatum(timeCtx, id, code, b)
+	_, err := DecodeOidDatum(context.Background(), timeCtx, id, code, b, nil)
 	if err != nil {
 		return 0
 	}

--- a/pkg/sql/pgwire/testdata/pgtest/enum
+++ b/pkg/sql/pgwire/testdata/pgtest/enum
@@ -11,6 +11,16 @@ ReadyForQuery
 {"Type":"CommandComplete","CommandTag":"DROP TYPE"}
 {"Type":"ReadyForQuery","TxStatus":"I"}
 
+send
+Query {"String": "DROP TABLE IF EXISTS tb"}
+----
+
+until ignore=NoticeResponse
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"DROP TABLE"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
 send crdb_only
 Query {"String": "SET experimental_enable_enums=true;"}
 ----
@@ -46,10 +56,10 @@ RowDescription
 # DataTypeSize for an enum is 4, as floats are used to represent enums
 # internally (4 bytes). Since our encodings are variable size, we report
 # the DataTypeSize to be -1, which is the variable length size.
-until ignore_type_oids crdb_only
+until crdb_only
 RowDescription
 ----
-{"Type":"RowDescription","Fields":[{"Name":"te","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":0,"DataTypeSize":-1,"TypeModifier":-1,"Format":0}]}
+{"Type":"RowDescription","Fields":[{"Name":"te","TableOID":0,"TableAttributeNumber":0,"DataTypeOID":100052,"DataTypeSize":-1,"TypeModifier":-1,"Format":0}]}
 
 until
 ReadyForQuery
@@ -57,3 +67,53 @@ ReadyForQuery
 {"Type":"DataRow","Values":[{"text":"hi"}]}
 {"Type":"CommandComplete","CommandTag":"SELECT 1"}
 {"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Regression for #53413. This test ensures that the wire protocol can handle
+# user defined type OIDs.
+
+send
+Query {"String": "CREATE TABLE tb (x te)"}
+----
+
+until
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"CREATE TABLE"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Prepare a query and type hint a user defined type. Then bind this prepared
+# statement with a user defined type argument ([104, 105] = 'hi').
+send
+Parse {"Name": "s1", "Query": "INSERT INTO tb VALUES ($1)", "ParameterOIDs": [100052]}
+Bind {"DestinationPortal": "p", "PreparedStatement": "s1", "ParameterFormatCodes": [0], "Parameters": [[104, 105]]}
+Execute {"Portal": "p"}
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"CommandComplete","CommandTag":"INSERT 0 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Ensure that our value was successfully inserted.
+send
+Query {"String": "SELECT * FROM tb"}
+----
+
+until ignore_table_oids ignore_type_oids noncrdb_only
+RowDescription
+----
+{"Type":"RowDescription","Fields":[{"Name":"x","TableOID":0,"TableAttributeNumber":1,"DataTypeOID":0,"DataTypeSize":-1,"TypeModifier":-1,"Format":0}]}
+
+until crdb_only
+RowDescription
+----
+{"Type":"RowDescription","Fields":[{"Name":"x","TableOID":54,"TableAttributeNumber":1,"DataTypeOID":100052,"DataTypeSize":-1,"TypeModifier":-1,"Format":0}]}
+
+until
+DataRow
+----
+{"Type":"DataRow","Values":[{"text":"hi"}]}

--- a/pkg/sql/pgwire/types_test.go
+++ b/pkg/sql/pgwire/types_test.go
@@ -136,7 +136,7 @@ func TestIntArrayRoundTrip(t *testing.T) {
 
 	b := buf.wrapped.Bytes()
 
-	got, err := pgwirebase.DecodeOidDatum(nil, oid.T__int8, pgwirebase.FormatText, b[4:])
+	got, err := pgwirebase.DecodeOidDatum(context.Background(), nil, oid.T__int8, pgwirebase.FormatText, b[4:], nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -215,7 +215,7 @@ func TestByteArrayRoundTrip(t *testing.T) {
 					b := buf.wrapped.Bytes()
 					t.Logf("encoded: %v (%q)", b, b)
 
-					got, err := pgwirebase.DecodeOidDatum(nil, oid.T_bytea, pgwirebase.FormatText, b[4:])
+					got, err := pgwirebase.DecodeOidDatum(context.Background(), nil, oid.T_bytea, pgwirebase.FormatText, b[4:], nil)
 					if err != nil {
 						t.Fatal(err)
 					}
@@ -486,7 +486,7 @@ func BenchmarkDecodeBinaryDecimal(b *testing.B) {
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
 		b.StartTimer()
-		got, err := pgwirebase.DecodeOidDatum(nil, oid.T_numeric, pgwirebase.FormatBinary, bytes)
+		got, err := pgwirebase.DecodeOidDatum(context.Background(), nil, oid.T_numeric, pgwirebase.FormatBinary, bytes, nil)
 		b.StopTimer()
 		evalCtx := tree.NewTestingEvalContext(cluster.MakeTestingClusterSettings())
 		defer evalCtx.Stop(context.Background())

--- a/pkg/sql/types/types.go
+++ b/pkg/sql/types/types.go
@@ -1219,8 +1219,14 @@ func RemapUserDefinedTypeOIDs(t *T, newOID, newArrayOID oid.Oid) {
 
 // UserDefined returns whether or not t is a user defined type.
 func (t *T) UserDefined() bool {
+	return IsOIDUserDefinedType(t.Oid())
+}
+
+// IsOIDUserDefinedType returns whether or not o corresponds to a user
+// defined type.
+func IsOIDUserDefinedType(o oid.Oid) bool {
 	// Types with OIDs larger than the predefined max are user defined.
-	return t.Oid() > oidext.CockroachPredefinedOIDMax
+	return o > oidext.CockroachPredefinedOIDMax
 }
 
 var familyNames = map[Family]string{

